pylhc/lsa_to_madx.py

```diff
@@ -0,0 +1,241 @@
+In this case many files will be created.
+The call would be:
+
+.. code-block:: bash
+
+    python -m pylhc.lsa_to_madx \\
+        --optics R2017aT_A30C30A10mL300_CTPPS2 \\
+        --file knobs.txt
+
+Hint: the knobs active at a given time can be retrieved with the `~pylhc.machine_settings_info` script. 
+"""
+import argparse
+
+from pathlib import Path
+from typing import Dict
+
+import tfs
+
+from omc3.utils import logging_tools
+from omc3.utils.contexts import timeit
+
+from pylhc.data_extract.lsa import LSAClient
+
+LOG = logging_tools.get_logger(__name__)
+
+# ----- Helper functions ----- #
+
+
+def parse_knobs_and_trim_values_from_file(knobs_file: Path) -> Dict[str, float]:
+    """
+    Parses a file for LSA knobs and their trim values. Each line should be a knob name
+    following by a number of the trim value. If no value is written, it will be defaulted
+    to ``1.0``. Lines starting with a ``#`` are ignored.
+
+    Args:
+        knobs_file (~pathlib.Path): Path to the file with definitions.
+
+    Returns:
+        A `dict` with as keys the parsed knob names and as values their associated trims.
+    """
+    knob_lines = [line for line in Path(knobs_file).read_text().splitlines() if not line.startswith("#")]
+    results = {}
+
+    for line in knob_lines:
+        knob = line.split()[0]  # this is the knob name
+        try:  # catch the trim value, which is a number after the knob name
+            trim = float(line.split()[1])
+            results[knob] = trim
+        except IndexError:  # the is no number, we default the trim value to 1.0
+            LOG.debug(f"No trim value was specified for knob '{knob}' - defaulting to 1.0")
+            results[knob] = 1.0
+    return results
+
+
+def get_madx_script_from_definition_dataframe(deltas_df: tfs.TfsDataFrame, lsa_knob: str, trim: float = 1.0) -> str:
+    """
+    Given the extracted definition dataframe of an LSA knob - as returned by
+    `~pylhc.data_extract.lsa.LSAClient.get_knob_circuits` - this function will generate the
+    corresponding ``MAD-X`` text commands that would reproduce this knob in a script.
+
+    Args:
+        deltas_df (~tfs.frame.TfsDataFrame): The extracted definition dataframe of an ``LSA`` knob. This
+            can be obtained with `~pylhc.data_extract.lsa.LSAClient.get_knob_circuits`.
+        lsa_knob (str): The complete ``LSA`` name of the knob, including any ``LHCBEAM/`` etc.
+        trim (float): The trim value to write for the knob. Defaults to 1.
+
+    Returns:
+        The complete ``MAD-X`` script to reproduce the knob, as a string.
+    """
+    LOG.debug(f"Determining MAD-X commands to reproduce knob '{lsa_knob}'")
+    change_commands = [f"! Start of change commands for knob: {lsa_knob}"]
+
+    # Set this to 1 by default but can be changed by the user to reproduce a given trim
+    knob_itself = lsa_knob.split("/")[1]  # without the LHCBEAM/ part
```

Comment:
if this is convention, maybe it's worth putting it as some sort of global regex str at the top

Reply:
Regex :(
As far as I know it is an almost convention, sometimes it is LHCBEAM1 or LHCBEAM2. I will see for a regex

Reply:
honestly, I would go with `lsa_knob.split("/")[-1]` . In that case it doesn't matter if there is a `/` in the name or not (only gets problematic when there is two, but that is even less likely.)